### PR TITLE
[part 1/2] Warp sync on light client

### PR DESCRIPTION
**Pull request by ordian**
Part of #6010.

This part doesn't include background old header sync.

### CODE REVIEW

File: ethcore/light/src/client/snapshot/chain.rs [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+	) {
+		let result = if header.number() == 0 {
+			let td = self.chain.genesis_header().difficulty();
+			self.chain.insert_with_td(batch, header, td, Some(transition.proof))
+		} else {
+			self.chain.insert(batch, header, Some(transition.proof))
+		};
+		match result {
+			Ok(pending) => *self.pending.write() = Some(pending),
+			// TODO: panic? abort snapshot sync?
```
Comment (dvdplm) [inline ethcore/light/src/client/snapshot/chain.rs]: Why not return error upstream and have?

```rust
fn insert_epoch_transition(… …) -> Result<…> {
  if header.number() == 0 {
    let td = self.chain.genesis_header().difficulty();
    self.chain.insert_with_td(batch, header, td, Some(transition.proof))
  } else {
    self.chain.insert(batch, header, Some(transition.proof))
  }
}
```

**EDIT** I see: the new trait wraps previous code that doesn't use `Return` much. Is it worthwhile to change that perhaps, so that new code can use more idiomatic error handling?

File: ethcore/light/src/client/mod.rs [hunk 2]
```diff
@@ -626,3 +648,19 @@ impl<T: ChainDataFetcher> ::ethcore::client::EngineClient for Client<T> {
 		Client::block_header(self, id)
 	}
 }
+
+impl<T: ChainDataFetcher> snapshot::DatabaseRestore for Client<T> {
+	/// Restart the client with a new backend
+	fn restore_db(&self, new_db: &str, spec: &Spec, hs: HardcodedSync) -> Result<(), Error> {
+		trace!(target: "snapshot", "Replacing light client database with {:?}", new_db);
+
+		let _import_lock = self.import_lock.lock();
```
[the remark this answers is not in the document]
Comment (ordian) [reply ethcore/light/src/client/mod.rs]: `try_lock_for()` is a timed lock. What should we do if the timeout is reached? If try again, then this is no better than `lock()`.

File: ethcore/light/src/client/header_chain.rs [hunk 3]
```diff
@@ -564,6 +510,67 @@ impl HeaderChain {
 		Ok(pending)
 	}
 
+	fn produce_next_cht_root(
+		&self, from_era: u64, candidates: &mut BTreeMap<u64, Entry>,
```
Comment (rphmeier) [inline ethcore/light/src/client/header_chain.rs]: style: each param should get its own line if there are too many for one line

File: ethcore/light/src/client/header_chain.rs [hunk 4]
```diff
@@ -701,6 +708,16 @@ impl HeaderChain {
 		}
 	}
 
+	/// Get the cache.
+	pub fn cache(&self) -> Arc<Mutex<Cache>> {
+		self.cache.clone()
+	}
+
+	/// Get the database column.
```
Comment (rphmeier) [inline ethcore/light/src/client/header_chain.rs]: would be good to elaborate on what this column is used for

File: ethcore/light/src/client/mod.rs [hunk 5]
```diff
@@ -515,14 +517,25 @@ impl<T: ChainDataFetcher> Client<T> {
...
 			proof: proof,
 		});
 		self.db.write_buffered(batch);
 		Ok(())
 	}
 }
 
+// An ugly borrow-checker workaround to allow iterating over a block and its ancestry.
+/// An iterator over a block and its ancestry.
+pub struct AncestryIter<'a>(RwLockReadGuard<'a, HeaderChain>);
```
Comment (rphmeier) [inline ethcore/light/src/client/mod.rs]: why is this necessary?

File: ethcore/src/engines/mod.rs [hunk 6]
```diff
@@ -342,9 +342,15 @@ pub trait Engine<M: Machine>: Sync + Send {
 		None
 	}
 
+	/// Create a factory for restoring from snapshot chunks.
+	/// Returning `None` indicates that this engine doesn't support restoring from a snapshot.
+	fn snapshot_restoration(&self) -> Option<Box<RebuilderFactory>> {
```
Comment (rphmeier) [inline ethcore/src/engines/mod.rs]: Should be unified with `SnapshotComponents`

File: ethcore/light/src/client/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (dvdplm) [reply ethcore/light/src/client/mod.rs]: Well, perhaps exit with an error message?

File: ethcore/light/src/client/snapshot/chain.rs [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+	fn block_hash(&self, index: BlockNumber) -> Option<H256> {
+		self.chain.block_hash(BlockId::Number(index))
+	}
+
+	fn block_header_data(&self, hash: &H256) -> Option<Header> {
+		self.chain.block_header(BlockId::Hash(hash.clone())).and_then(|h| h.decode().ok())
+	}
+
+	fn add_child(&self, _batch: &mut DBTransaction, _block_hash: H256, _child_hash: H256) {
+		// TODO
```
Comment (ordian) [inline ethcore/light/src/client/snapshot/chain.rs]: Should we store this parent<->child relationship in the database similar to a full node 
https://github.com/paritytech/parity/blob/0bed5976e368e948566f5a445f5fa17088674f3d/ethcore/src/blockchain/blockchain.rs#L934-L947?

File: ethcore/light/src/client/header_chain.rs [hunk 8]
```diff
@@ -462,11 +464,15 @@ impl HeaderChain {
...
+				trace!(
+					target: "chain", "Setting new canonical block {} for block height {}",
+					canon_hash, height
+				);
 
-				let canon_pos = entry.candidates.iter().position(|x| x.hash == canon_hash)
-					.expect("blocks are only inserted if parent is present; or this is the block we just added; qed");
+				let canon_pos = match entry.candidates.iter().position(|x| x.hash == canon_hash) {
+					Some(pos) => pos,
+					None => break,
```
Comment (ordian) [inline ethcore/light/src/client/header_chain.rs]: Not sure about correctness here.

**Comment by 5chdn** [comment]
Is this to be reviewed? "[WIP] [1/2]"?

**Comment by 5chdn** [comment]
> Is this to be reviewed? "[WIP] [1/2]"?

@ordian :)

File: ethcore/light/Cargo.toml [hunk 9]
```diff
@@ -36,10 +36,13 @@ triehash = { path = "../../util/triehash" }
 kvdb = { path = "../../util/kvdb" }
 memory-cache = { path = "../../util/memory_cache" }
 error-chain = { version = "0.12", default-features = false }
+util-error = { path = "../../util/error" }
```
Comment (niklasad1) [inline ethcore/light/Cargo.toml]: This was crate was removed https://github.com/paritytech/parity/pull/9054, do you need it?

Please rebase/merge to master

File: ethcore/src/snapshot/consensus/mod.rs [hunk 10]
```diff
@@ -51,17 +54,54 @@ pub trait SnapshotComponents: Send {
 		chunk_sink: &mut ChunkSink,
 		preferred_size: usize,
 	) -> Result<(), Error>;
+}
+
+/// A common trait for `BlockChain` (full node) and `HeaderChain` (light client)
+/// used in snapshot restoration.
+pub trait RestorationTargetChain: Send {
+	/// Same as `BlockChain::genesis_hash`.
```
Comment (dvdplm) [inline ethcore/src/snapshot/consensus/mod.rs]: Perhaps move doc comments from `BlockChain` to this trait?

File: ethcore/src/snapshot/consensus/mod.rs [hunk 11]
```diff
@@ -51,17 +54,54 @@ pub trait SnapshotComponents: Send {
...
+	fn genesis_header(&self) -> Header;
+
+	/// Same as `BlockChain::block_hash`.
+	fn block_hash(&self, index: BlockNumber) -> Option<H256>;
+
+	/// Same as `BlockChain::block_header_data`.
+	fn block_header_data(&self, hash: &H256) -> Option<Header>;
+
+	/// Same as `BlockChain::add_child`.
+	fn add_child(&self, batch: &mut DBTransaction, block_hash: H256, child_hash: H256);
```
Comment (dvdplm) [inline ethcore/src/snapshot/consensus/mod.rs]: If the LightClient doesn't need this – and I don't have an answer to that I'm afraid – perhaps this doesn't need to be in the trait and can be moved back to the `BlockChain` impl?

File: ethcore/src/snapshot/consensus/mod.rs [hunk 12]
```diff
@@ -51,17 +54,54 @@ pub trait SnapshotComponents: Send {
...
+	fn insert_unordered_block(
+		&self, batch: &mut DBTransaction, bytes: &[u8], receipts: Vec<Receipt>,
+		parent_td: Option<U256>, is_best: bool, is_ancient: bool
+	) -> bool;
+
+	/// Same as `BlockChain::commit`
+	fn commit(&self);
+}
+
+/// A factory producing `Rebuilder` necessary for snapshot restoration.
```
Comment (dvdplm) [inline ethcore/src/snapshot/consensus/mod.rs]: s/producing `Rebuilder`/producing a `Rebuilder`/

File: ethcore/src/snapshot/consensus/mod.rs [hunk 13]
```diff
@@ -51,17 +54,54 @@ pub trait SnapshotComponents: Send {
...
+pub trait RebuilderFactory: Send {
+	/// Create a rebuilder, which will have chunks fed into it in arbitrary
```
Comment (dvdplm) [inline ethcore/src/snapshot/consensus/mod.rs]: s/rebuilder/`Rebuilder`/

File: ethcore/sync/src/light_sync/warp_sync.rs [hunk 14]
```diff
@@ -0,0 +1,652 @@
...
+		self.peers.len()
+	}
+
+	/// Get peer by id.
+	pub fn get(&self, peer: &PeerId) -> Option<&SnapshotPeer> {
+		self.peers.get(peer)
+	}
+
+	/// Returns the largest group of peers with the same snapshot hash.
+	/// If we already  have manifest, prefer peers with that manifest.
```
Comment (dvdplm) [inline ethcore/sync/src/light_sync/warp_sync.rs]: Double spaces between "already" and "have"

File: ethcore/sync/src/light_sync/warp_sync.rs [hunk 15]
```diff
@@ -0,0 +1,652 @@
...
+
+		if self.peers.contains_key(&peer_id) {
+			debug!(target: "warp", "Unexpected status packet from a known peer {}", peer_id);
+			return Ok(());
+		}
+
+		let protocol_version: u8 = rlp.val_at(0)?;
+		let network_id: u64 = rlp.val_at(1)?;
+		let _difficulty: U256 = rlp.val_at(2)?;
+		let _latest_hash: H256 = rlp.val_at(3)?;
```
Comment (dvdplm) [inline ethcore/sync/src/light_sync/warp_sync.rs]: Why fetch these two if we don't need them? (and why bail on error?) Is it faster maybe?

File: parity/export_hardcoded_sync.rs [hunk 16]
```diff
@@ -86,14 +86,40 @@ pub fn execute(cmd: ExportHsyncCmd) -> Result<String, String> {
...
+	let client_config = to_client_config(
+		&cmd.cache_config,
+		spec.name.to_lowercase(),
+		Mode::Active,
+		tracing,
+		fat_db,
+		cmd.compaction,
+		cmd.wal,
+		VMType::Interpreter,
+		"".into(),
```
Comment (dvdplm) [inline parity/export_hardcoded_sync.rs]: Perhaps annotate these hard-to-guess items to help future readers?

File: ethcore/light/src/client/mod.rs [hunk 17]
```diff
@@ -159,7 +161,7 @@ impl<T: LightChainClient> AsLightClient for T {
 pub struct Client<T> {
 	queue: HeaderQueue,
 	engine: Arc<EthEngine>,
-	chain: HeaderChain,
+	chain: RwLock<HeaderChain>,
```
[the remark this answers is not in the document]
Comment (ordian) [reply ethcore/light/src/client/mod.rs]: `HeaderChain` is thread-safe (it's fields are wrapped in mutexes/rwlocks) itself, but we need to replace `chain` field in a thread safe way [here](https://github.com/paritytech/parity/pull/8723/files#diff-223b03b977353484865ff436905fac0cR656).

File: ethcore/sync/src/light_sync/warp_sync.rs (hunk 15 above)
Comment (ordian) [reply ethcore/sync/src/light_sync/warp_sync.rs]: To ensure that the status message that we receive from a full node is well formed.

File: ethcore/light/src/client/mod.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (dvdplm) [reply ethcore/light/src/client/mod.rs]: ah gotcha, makes sense.

File: ethcore/light/src/client/mod.rs [hunk 18]
```diff
@@ -514,14 +517,28 @@ impl<T: ChainDataFetcher> Client<T> {
...
+// An ugly borrow-checker workaround to allow iterating over a block and its ancestry.
+// We need this since `self.chain.read()` does not live long enough for
+// `self.chain.read().ancestry_iter(_)` to work. And we probably don't want to expose
+// `RwLockReadGuard` in the public API.
+/// An iterator over a block and its ancestry.
+pub struct AncestryIter<'a>(RwLockReadGuard<'a, HeaderChain>);
+
+impl<'a> AncestryIter<'a> {
+	/// Get an iterator over a block and its ancestry.
+	pub fn iter(&'a self, start: BlockId) -> Box<Iterator<Item=encoded::Header> + 'a> {
```
Comment (rphmeier) [inline ethcore/light/src/client/mod.rs]: could even be `impl Iterator` now

File: ethcore/light/src/client/service.rs [hunk 19]
```diff
@@ -98,14 +135,46 @@ impl<T: ChainDataFetcher> Service<T> {
...
+/// IO interface for the Client handler
+struct ClientIoHandler<T> {
+	client: Arc<Client<T>>,
+	snapshot: Arc<SnapshotService>,
 }
 
-struct ImportBlocks<T>(Arc<Client<T>>);
+impl<T> ClientIoHandler<T> {
+	pub fn new(client: Arc<Client<T>>, snapshot: Arc<SnapshotService>) -> Self {
+		Self {
```
Comment (rphmeier) [inline ethcore/light/src/client/service.rs]: favor `ClientIoHandler`

File: ethcore/light/src/client/snapshot/chain.rs (hunk 7 above)
Comment (rphmeier) [reply ethcore/light/src/client/snapshot/chain.rs]: yes, something like that

**Comment by 5chdn** [comment]
Is CI still not working?

File: ethcore/sync/src/chain/mod.rs [hunk 20]
```diff
@@ -155,11 +155,11 @@ const MAX_TRANSACTION_PACKET_SIZE: usize = 8 * 1024 * 1024;
...
 const MAX_TRANSACTIONS_TO_PROPAGATE: usize = 64;
-// Min number of blocks to be behind for a snapshot sync
-const SNAPSHOT_RESTORE_THRESHOLD: BlockNumber = 30000;
-const SNAPSHOT_MIN_PEERS: usize = 3;
+/// Min number of blocks to be behind for a snapshot sync
+pub const SNAPSHOT_RESTORE_THRESHOLD: BlockNumber = 30000;
+pub const SNAPSHOT_MIN_PEERS: usize = 3;
 
-const STATUS_PACKET: u8 = 0x00;
+pub const STATUS_PACKET: u8 = 0x00;
```
Comment (rphmeier) [inline ethcore/sync/src/chain/mod.rs]: document all the public constants. Also, why have these become public?

File: ethcore/sync/src/light_sync/mod.rs [hunk 21]
```diff
@@ -275,6 +300,13 @@ impl<L: AsLightClient + Send + Sync> Handler for LightSync<L> {
 	}
 
 	fn on_disconnect(&self, ctx: &EventContext, unfulfilled: &[ReqId]) {
+		match *self.state.lock() {
+			SyncState::Snapshot(_) => {
+				return;
```
Comment (rphmeier) [inline ethcore/sync/src/light_sync/mod.rs]: extraneous braces

File: ethcore/sync/src/light_sync/mod.rs [hunk 22]
```diff
@@ -275,6 +300,13 @@ impl<L: AsLightClient + Send + Sync> Handler for LightSync<L> {
 	}
 
 	fn on_disconnect(&self, ctx: &EventContext, unfulfilled: &[ReqId]) {
+		match *self.state.lock() {
```
Comment (rphmeier) [inline ethcore/sync/src/light_sync/mod.rs]: could be an `if let`, right?

Comment (ordian) [reply ethcore/sync/src/light_sync/mod.rs]: No, there is a slight difference due to borrowing rules: [play link](http://play.rust-lang.org/?gist=b1d4378f8e0f0ac1a02028365e24b524&version=stable&mode=debug&edition=2015).

File: ethcore/sync/src/chain/mod.rs (hunk 20 above)
Comment (ordian) [reply ethcore/sync/src/chain/mod.rs]: Because currently, the warp sync handshake packet uses it: https://github.com/paritytech/parity-ethereum/blob/7ad556346eed03cfea2999f7f9daf864de6489af/ethcore/sync/src/chain/mod.rs#L890-L910

File: ethcore/light/src/client/header_chain.rs [hunk 23]
```diff
@@ -451,15 +453,23 @@ impl HeaderChain {
 		// reorganize ancestors so canonical entries are first in their
 		// respective candidates vectors.
 		if is_new_best {
+			let prove = "blocks are only inserted if parent is present; or this is the block we just added; qed";
```
Comment (dvdplm) [inline ethcore/light/src/client/header_chain.rs]: make `const` (and "proof" is a better name imo)

**Comment by 5chdn** [comment]
Has some conflicts and needs a 2nd review.

**Comment by 5chdn** [comment]
What's the status of this?

**Comment by ordian** [comment]
@5chdn the current implementation uses existing warp sync protocol (`WARP_SYNC_PROTOCOL_ID`), because it required less changes to implement. But because this protocol is tightly coupled with `eth` sync protocol, this approach has many drawbacks. 
I started working on switching to a separate protocol, but it's not in a reviewable state yet, since I focused on higher priority light client bugs and also `EIP#1218` may render it obsolete IIUC.

**Comment by 5chdn** [comment]
Ok, let's just reopen this once we are continuing with this.
